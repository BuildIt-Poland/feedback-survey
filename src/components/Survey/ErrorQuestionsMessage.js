import React from 'react';
import { string } from 'prop-types';
import { Box, Heading } from 'grommet';

import { colorRed } from '../../styles/variables';

const ErrorQuestionsMessage = ({ message }) => (
  <Box align="center" justify="center" pad="medium" height="90vh">
    <Heading level="4" textAlign="center" color={colorRed}>
      {message}
    </Heading>
  </Box>
);

ErrorQuestionsMessage.propTypes = {
  message: string.isRequired
};

export default ErrorQuestionsMessage;
