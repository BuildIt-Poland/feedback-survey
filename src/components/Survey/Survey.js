import React from 'react';
import { Redirect } from 'react-router-dom';

import { Form, Box, Heading } from 'grommet';

import { fetchQuestions, saveSurvey } from '../../services/api';
import SurveyField from '../SurveyField/SurveyField';
import SurveyButton from '../Button';

const messages = {
  required: 'This field is required'
};

class Survey extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: props.isLoading || true,
      data: { questions: [], answerTypes: {} },
      submitted: false,
      success: false,
      error: null
    };
  }

  componentDidMount() {
    if (this.state.isLoading) {
      fetchQuestions().then(res => {
        this.setState({
          data: res.data,
          isLoading: false
        });
      });
    }
  }

  renderSurvey() {
    const { submitted, success, error, data, isLoading } = this.state;
    if (submitted) {
      return (
        <Redirect
          to={{
            pathname: '/final',
            state: { success, error }
          }}
        />
      );
    } else {
      if (isLoading) {
        return (
          <Box animation="pulse" align="center" justify="center" pad="medium" height="90vh">
            <Heading level="4">Loading...</Heading>
          </Box>
        );
      } else {
        return (
          <Box align="center" justify="center" pad="medium">
            <Form onSubmit={this.handleSubmit} messages={messages}>
              {this.renderFormFields(data)}
              <Box align="center">
                <SurveyButton type="submit" label="submit" margin="xlarge" data-testid="survey-submit-button" />
              </Box>
            </Form>
          </Box>
        );
      }
    }
  }

  renderFormFields(data) {
    return data.questions.map((field, index) => (
      <SurveyField key={index} field={field} answerTypes={data.answerTypes} />
    ));
  }

  handleSuccess = () => {
    this.setState({
      submitted: true,
      success: true
    });
  };

  handleError = error => {
    this.setState({
      submitted: true,
      error: error.message
    });
  };

  handleSubmit = event => {
    const data = {
      surveyId: this.props.match.params.surveyId || '',
      answers: event.value
    };
    saveSurvey(data).then(res => this.handleSuccess(), err => this.handleError(err));
  };

  render() {
    return this.renderSurvey();
  }
}

export default Survey;
