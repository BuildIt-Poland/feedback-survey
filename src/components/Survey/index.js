import { withRouter } from 'react-router-dom';

import Survey from './Survey';

export default withRouter(Survey);
