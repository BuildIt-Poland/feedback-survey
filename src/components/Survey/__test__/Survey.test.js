import React from 'react';
import { create } from 'react-test-renderer';

import Survey from '../Survey';

describe('COMPONENT - Survey', () => {
  it('renders correctly', () => {
    const component = create(<Survey />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
