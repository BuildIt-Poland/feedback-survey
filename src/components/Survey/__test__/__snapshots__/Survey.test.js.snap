// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - Survey renders correctly 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 hUosMD"
>
  <h4
    className="StyledHeading-sc-1rdh4aw-0 hlZfma"
  >
    Loading...
  </h4>
</div>
`;
